[GStreamer][MSE] Assign track indices sequentially by track type

vivienne-w opened the pull request:
#### c0ef14244e076bfa3c9ed980bb9f84d4f853906a
<pre>
[GStreamer][MSE] Assign track indices sequentially by track type
<a href="https://bugs.webkit.org/show_bug.cgi?id=287954">https://bugs.webkit.org/show_bug.cgi?id=287954</a>

Reviewed by Alicia Boya Garcia.

In other words: The first track of a type will always have index 0,
the next of the same type will be assigned 1, and so on.

This:
- matches how MediaPlayerPrivateGStreamer assigns indices
- how {Audio,Video,Text}TrackList stores its tracks
- prevents a potential out-of-bounds crash with inband text tracks, once we support those in MSE.

* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:

Canonical link: <a href="https://commits.webkit.org/290740@main">https://commits.webkit.org/290740@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2410d1b0f0563ccce0beb3fe57245c9fea453518

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90636 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10168 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45576 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95648 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41418 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92690 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10566 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18487 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69758 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27306 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93637 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8072 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82203 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50099 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7823 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36585 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40548 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78142 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37646 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97477 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17828 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13111 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78781 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18086 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78025 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77973 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/19283 "The change is no longer eligible for processing.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22413 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/21041 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14350 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17838 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23184 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17577 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21033 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19361 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -796,16 +796,21 @@ std::pair<AppendPipeline::CreateTrackResult, AppendPipeline::Track*> AppendPipel
     }
 
     size_t newTrackIndex = m_tracks.size();
+    size_t webkitTrackIndex = std::count_if(m_tracks.begin(), m_tracks.end(), [streamType](std::unique_ptr<Track>& other) -> bool {
```
This variable is only used in the `GST_DEBUG_OBJECT()` below, is that intentional or did you miss something?

vivienne-w replied:
I did miss that, thank you!

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -796,18 +796,23 @@ std::pair<AppendPipeline::CreateTrackResult, AppendPipeline::Track*> AppendPipel
+
     track.initializeElements(this, GST_BIN(m_pipeline.get()));
-    track.webKitTrack = makeWebKitTrack(newTrackIndex, trackId);
+    track.webKitTrack = makeWebKitTrack(webkitTrackIndex, trackId);
```
This will fail inside `makeWebKitTrack()`, as the first line of that function is:

```
Track& appendPipelineTrack = *m_tracks.at(trackIndex);
```

I'm also not sure it makes sense to use the indices of the `AppendPipeline::Track` in `m_tracks` for anything outside of `AppendPipeline`, but I can also confirm by looking at 569fdc337c908665da86de50acf2adcf89d7624e that it is my fault, as I was already doing that in the original multi-track support patch.

vivienne-w replied:
Wouldn't the most straightforward solution here be to just pass the `Track&` as another argument to `makeWebKitTrack()`?

I'm also not sure what you mean by using the indices outside `AppendPipeline`?

ntrrgc replied:
Assuming nothing else is broken, yes.

> I'm also not sure what you mean by using the indices outside AppendPipeline?

I mean how the `trackIndex` argument of `makeWebKitTrack()` is passed to the constructor of VideoTrackPrivateGStreamer etc.

ntrrgc replied:
FWIW I was able to confirm the indices for TextTrackPrivateGStreamer and TextTrackPrivateBase are both the same, and per-type.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -796,18 +796,23 @@ std::pair<AppendPipeline::CreateTrackResult, AppendPipeline::Track*> AppendPipel
 
     size_t newTrackIndex = m_tracks.size();
+    size_t webkitTrackIndex = std::count_if(m_tracks.begin(), m_tracks.end(), [streamType](std::unique_ptr<Track>& other) -> bool {
+        return other->streamType == streamType;
```
The problem I see with this is that the indices would still be per-AppendPipeline (and by extension per-SourceBuffer), but if I understood things correctly now, the indices in TrackPrivateGStreamer should be per-MediaPlayer/MediaSource.

However -- again assuming I got it right -- this is not a new problem since the original multi-track patch already had that flaw, so it feels unfair to request you fix that here now.

vivienne-w replied:
I'll keep this in mind as a potential task in the future, but I think for now I think this works well enough.

vivienne-w replied:
I've implemented this here: https://github.com/WebKit/WebKit/pull/42200

ntrrgc reviewed (approved):
I'll leave it to your judgement whether to try to fix the tangentially related issue I commented in this PR or to leave it for another patch.

vivienne-w committed:
[GStreamer][MSE] Assign track indices sequentially by track type
https://bugs.webkit.org/show_bug.cgi?id=287954

Reviewed by Alicia Boya Garcia.

In other words: The first track of a type will always have index 0,
the next of the same type will be assigned 1, and so on.

This:
- matches how MediaPlayerPrivateGStreamer assigns indices
- how {Audio,Video,Text}TrackList stores its tracks
- prevents a potential out-of-bounds crash with inband text tracks, once we support those in MSE.

* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:

Canonical link: https://commits.webkit.org/290740@main
